src/main.rs

```diff
@@ -1,767 +0,0 @@
-        inc_metric!(json_rpc_request_err_no_permission);
-        return Err(EthRpcError::NoPermission);
-    }
-    let cycles_available = ic_cdk::api::call::msg_cycles_available128();
-    let parsed_url = url::Url::parse(&service_url).or(Err(EthRpcError::ServiceUrlParseError))?;
-    let host = parsed_url
-        .host_str()
-        .ok_or(EthRpcError::ServiceUrlHostMissing)?
-        .to_string();
-    if SERVICE_HOSTS_ALLOWLIST.with(|a| !a.borrow().contains(&host.as_str())) {
-        log!(INFO, "host not allowed {}", host);
-        inc_metric!(json_rpc_request_err_service_url_host_not_allowed);
-        return Err(EthRpcError::ServiceUrlHostNotAllowed);
-    }
-    if !authorized(Auth::FreeRpc) {
-        let provider_cost = match &provider {
-            None => 0,
-            Some(provider) => json_rpc_provider_cycles_cost(
-                &json_rpc_payload,
-                provider.cycles_per_call,
-                provider.cycles_per_message_byte,
-            ),
-        };
-        let cost = json_rpc_cycles_cost(&json_rpc_payload, &service_url, max_response_bytes)
-            + provider_cost;
-        if cycles_available < cost {
-            return Err(EthRpcError::TooFewCycles(format!(
-                "requires {} cycles, got {} cycles",
-                cost, cycles_available
-            )));
-        }
-        ic_cdk::api::call::msg_cycles_accept128(cost);
-        if let Some(mut provider) = provider {
-            provider.cycles_owed += provider_cost;
-            PROVIDERS.with(|p| {
-                // Error should not happen here as it was checked before.
-                p.borrow_mut()
-                    .insert(provider.provider_id, provider)
-                    .expect("unable to update Provider");
-            });
-        }
-        add_metric!(json_rpc_request_cycles_charged, cost);
-        add_metric!(json_rpc_request_cycles_refunded, cycles_available - cost);
-    }
-    inc_metric_entry!(json_rpc_host_requests, host);
-    let request_headers = vec![
-        HttpHeader {
-            name: "Content-Type".to_string(),
-            value: "application/json".to_string(),
-        },
-        HttpHeader {
-            name: "Host".to_string(),
-            value: host.to_string(),
-        },
-    ];
-    let request = CanisterHttpRequestArgument {
-        url: service_url,
-        max_response_bytes: Some(max_response_bytes),
-        method: HttpMethod::POST,
-        headers: request_headers,
-        body: Some(json_rpc_payload.as_bytes().to_vec()),
-        transform: Some(TransformContext::new(transform, vec![])),
-    };
-    match make_http_request(request).await {
-        Ok((result,)) => Ok(result.body),
-        Err((r, m)) => {
-            inc_metric!(json_rpc_request_err_http_request_error);
-            Err(EthRpcError::HttpRequestError {
-                code: r as u32,
-                message: m,
-            })
-        }
-    }
-}
-
-fn json_rpc_cycles_cost(
```

Comment:
Note that this cost is multiplied by the following factors `subnet_size/13`

Reply:
Added a constant for subnet size.